sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/MainSafeCoroutinesCheck.kt

```diff
@@ -0,0 +1,154 @@
+        val bindingContext = context.bindingContext
+        val suspendModifier = function.suspendModifier()
+        if (suspendModifier != null) {
+            function.reportBlockingFunctionCalls(context)
+        } else {
+            function.forEachDescendantOfType<KtLambdaArgument> {
+                if (it.isSuspending(bindingContext)) it.reportBlockingFunctionCalls(context)
+            }
+        }
+    }
+
+    private fun KtElement.reportBlockingFunctionCalls(context: KotlinFileContext) {
+        val bindingContext = context.bindingContext
+        forEachDescendantOfType<KtCallExpression> { call ->
+            val resolvedCall = call.getResolvedCallNative(bindingContext)
+            if (resolvedCall matches THREAD_SLEEP_MATCHER) {
+                context.reportIssue(call.calleeExpression!!,
+                    """Replace this "Thread.sleep()" call with "delay()"""")
+            } else {
+                resolvedCall?.resultingDescriptor?.let { descriptor ->
+                    if (isInsideNonSafeDispatcher(call,
+                            bindingContext) && descriptor.annotations.any { throwsBlockingExceptions(it) }
+                    ) {
+                        context.reportIssue(call.calleeExpression!!,
+                            """Use "Dispatchers.IO" to run this potentially blocking operation""")
+                    }
+                }
+            }
+        }
+    }
+}
+
+private fun isInsideNonSafeDispatcher(
+    callExpr: KtCallExpression,
+    bindingContext: BindingContext,
+): Boolean {
+    var parentCall: Call? = callExpr.getParentCall(bindingContext) ?: return true
+    var resolvedCall = parentCall.getResolvedCallNative(bindingContext) ?: return false
+
+    while (!FUNS_ACCEPTING_DISPATCHERS.any { resolvedCall matches it }) {
+        parentCall = parentCall?.callElement?.getParentCall(bindingContext)
+        if (parentCall == null) return true
+        val newResolvedCall = parentCall.getResolvedCallNative(bindingContext)
+        if (newResolvedCall === resolvedCall || newResolvedCall == null) return false
+        resolvedCall = newResolvedCall
+    }
+
+    return resolvedCall.usesNonSafeDispatcher(bindingContext)
+}
+
+private fun ResolvedCall<*>.usesNonSafeDispatcher(bindingContext: BindingContext): Boolean {
+    val arg = valueArgumentsByIndex?.get(0)
+    val argValue = ((arg as? ExpressionValueArgument)
+        ?.valueArgument
+        ?.getArgumentExpression()
+        ?.predictRuntimeValueExpression(bindingContext) as? KtDotQualifiedExpression)
+        ?.resolveReferenceTarget(bindingContext)
+        ?.fqNameOrNull()
+        ?.asString()
+
+    return arg == null
+        || arg is DefaultValueArgument
+        || argValue == "$KOTLINX_COROUTINES_PACKAGE.Dispatchers.Main"
+        || argValue == "$KOTLINX_COROUTINES_PACKAGE.Dispatchers.Default"
+}
+
+private fun throwsBlockingExceptions(annotation: AnnotationDescriptor) =
```

Comment:
Could this be valuable as function in `ApiExtensions`? Which takes e.g. a collection of exceptions that it checks for? That way we can centrally update it as well, once we find / the compiler provides us with a better way to detect which exceptions may be thrown.